[WIP] Introduce a new tool to check log output of a command.

pchan opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This PR adds a new tool that validates the log output of a command. The user has to specify the command and the expected output. This tool is intended to add integration tests for logging component (k8s.io/component-base/logs), but its usage is intended to more generic. The README.md has an elaborate explanation on the tool internals and usage. This is a WIP and is intended to get comments. 

#### Which issue(s) this PR fixes:
Fixes #102237

#### Special notes for your reviewer:
The following are some of the opens:
1. Comparing the output of the command needs more analysis/work. (need to consider stdout/stderr)
2. Should we use the Finite State Machine or markdown parser.
3. I am currently using fmt.Println for output as it's in RFC, I will move to use standardized APIs.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```
/wg structured-logging
/area logging
/sig testing
/cc @serathius @pohly

pchan committed:
Introduce a new tool to check log output of a command.

shivanshu1333 commented:
/ok-to-test

pohly commented on hack/verify-checkoutput.sh:
```diff
@@ -0,0 +1,41 @@
+# limitations under the License.
+
+# This script does a fast type check of kubernetes code for all platforms.
+# Usage: `hack/verify-typecheck.sh`.
```
This description is consistent with my expectation: without parameters, the script will check *all* binaries for which checking is enabled. However, the implementation doesn't match that behavior (yet).

To implement it, we need to define what "all binaries" and "checking is enabled" means. I think the answers are:
- find all directories with a "main" package
- there are one or more .md files inside that directory which contain test cases

pchan replied:
Ok couple of questions. Do we expect a coupling between the .md file and the main package ? If so, we need to validate that the test case corresponds to package main. How about we initially only target README.md and main.go ? I see a number of such instances that would be caught by this restriction [1].

[1] 
staging/src/k8s.io/component-base/logs/kube-log-runner
 staging/src/k8s.io/client-go/examples/dynamic-create-update-delete-deployment/

pohly replied:
You mean that you want to look for a `main.go` file instead of `main` package? As you noticed, Go does not require that such a file exists. I think it is better to to look for `^package main$` in .go files, which should be reasonably fast and pretty accurate.

I don't know whether a package may be called `main` without implementing a `main` function, i.e. not implement a command. Even if it is allowed, it would be so confusing that no-one is doing it.

pchan replied:
I have updated the code and the README.md to reflect usage of main package and the markdown file.

pohly commented on test/checkoutput/README.md:
````diff
@@ -0,0 +1,54 @@
+
+**Usage:**
+```
+<tool> -parsefile <path/to/file> -verbose
````
Instead of `-verbose`, perhaps enable logging options?

I think I prefer `<tool> <directory> <directory> ...` as invocation. Then the tool can check the binary defined in each directory based on the files defined there.

We can start with test cases embedded in .md files. Supporting comments in .go files could be added later, if we find a need for it.

pchan replied:
I don't understand the alternative for -verbose. What does it mean to enable logging options ? Can you please elaborate.

pohly replied:
Enable `-v` and then in the tool, use `klog.V(...).InfoS` to print information at various level of detail.

This can be done as in https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/component-base/logs/example

pohly commented on test/checkoutput/README.md:
```diff
@@ -0,0 +1,54 @@
+1. Parsing and capturing the Run/Verify blocks
+2. Run and verify the output of the blocks captured in the previous step
+
+## Parsing
```
This belongs into source code comments, not the end-user documentation (= README.md). What we need here is a definition of how a test case gets defined and in which environment it gets executed.

For example, what is the current directory? I think it should be the directory that is being tested. That's consistent with "go test" and enables referencing local files without a long path that might change when code gets moved around.

Also, `go run ./staging/....` doesn't work in k/k. One has to be inside the module defined inside `staging`.

pchan replied:
I will move the parsing part into the go file. We will move to a directory structure instead of specify files, and that should serve as the current directory. I will change the code to reflect it.

pohly commented on test/checkoutput/README.md:
````diff
@@ -0,0 +1,54 @@
+
+Run:
+```console
+go run ./staging/src/k8s.io/component-base/logs/example/cmd/logger.go
````
This works only because `logger.go` is self-contained. It does not work for binaries which are split into different files, because then the entire package has to be compiled and that doesn't work for `staging` (see comment above):
```
$ go run ./staging/src/k8s.io/component-base/logs/example/cmd
go run: no packages loaded from ./staging/src/k8s.io/component-base/logs/example/cmd
```

`go run .` works if we invoke commands inside the binary's directory.

pohly replied:
I've argued before that pre-built binaries should be used. I still see some value in that (for example, it ensures that binaries are tested exactly as they will get released), but perhaps "go run" is close enough. Let's keep it simple for now and use "go run".

pchan replied:
Do you have any examples of pre-built binaries that I can use in the tool ?

pohly replied:
Just run "make" and check under `_output`. But this isn't important right now. `go run .` will work fine for most cases, so let's start with that.

pohly commented on test/checkoutput/README.md:
```diff
@@ -0,0 +1,54 @@
+
+# Open Issues
+## stdout and stderr
+Currently we only check validate stdout. We need to come up with a proper mechanism to validate outputs in stdout and stderr as this tool is intended to check logging.
```
As a first step, the tool can capture both stdout and stderr in a single stream. Then different test cases can redirect stdout or stderr to /dev/null to test just one output stream in isolation.

But I agree that this makes writing tests cases more verbose. I agree that it would be nice to support testing "Expected stderr" and "Expected stdout" with separate streams. If one or the other is not specified, the output has to be empty.

pchan replied:
I am currently using a combined output (both stdout and stderr). We will revisit this to see if we need to have a separate streams.

pohly reviewed (changes requested):
I've only looked at the usage. Let's agree on the desired behavior first.

pchan commented:
I had rebased, so I was forced to do a force-push. I have incorporated the feedback. The pending items are 

1. Need more examples of pre-built binaries.
2. The hack/verify-checkoutput seems to set the GOPATH to k8s build/output and so `go run .` in staging/src/k8s.io/component-base/logs/example/cmd/ does not work. Need to investigate this. Manually setting GOPATH and running the tool works.
3. Move from fmt.Print* to klogs.

serathius commented on test/checkoutput/sample_input.md:
````diff
@@ -0,0 +1,80 @@
+
+Run:
+```console
+go run /home/prasadc/dev/k8s/repo/k8s/k1/src/k8s.io/kubernetes/staging/src/k8s.io/component-base/logs/example/cmd/logger.go --logging-format json
````
Please remove your local path.

pchan replied:
Done.

pchan commented:
I have added klogs, updated the README and the logger.go in component-base example in the latest commit.

pohly commented on hack/verify-checkoutput.sh:
```diff
@@ -0,0 +1,41 @@
+hack/run-in-gopath.sh \
+    go run test/checkoutput/main.go "$@" || ret=$?
+if [[ $ret -ne 0 ]]; then
+  echo "!!! checkoutput has failed. This mostly implies a parsing issue" >&2
```
Why does that imply a parsing issue?

We want a non-zero return code for both invalid input and test failures.

pchan replied:
This comment was made when it originally took a single Markdown file as input. I forgot to update it.

pchan replied:
Fixed.

pohly commented on test/checkoutput/README.md:
```diff
@@ -0,0 +1,45 @@
+# checkoutput tool
+
+The checkoutput tool parses a text or markdown that has a command and the expected output sequence, runs the command and checks the output of the command against the expected output. 
```
markdown is spelled Markdown (capital letter).

pchan replied:
Fixed.

pohly commented on test/checkoutput/README.md:
````diff
@@ -0,0 +1,45 @@
+```
+Expected output:
+```
+I0605 22:03:07.224293 3228948 logger.go:58] Log using Infof, key: value
````
This example demonstrates a key problem: the time stamps will never match.

pohly replied:
We have to support plain text matching and regular expression matching.

pchan replied:
Yes, I noticed that too when I ran it. I will add regular expression matching.

pohly commented on test/checkoutput/main.go:
```diff
@@ -0,0 +1,301 @@
+		return nil
+	}
+	// This implies a validation failure
+	fmt.Printf("Validation Failure: command : %v \n Got :\n%s\n Expected %v\n", rv.run, string(output), rv.verify)
```
No-one is going to read the log of a CI job to find such `Validation Failure` messages. If there are validation errors, the command must return a non-zero return code to indicate to the caller that something went wrong.

pchan replied:
So do we want the failure to be flagged immediately and terminate the run ? 

In the README I had proposed multiple failures being captured in a single run. Can we use a flag to toggle the behavior ?

> ## Failure                                                                                                                                                                             
>                                                                                                                                                                                        
> Multiple Run/Verify blocks can exist and a failure is treated as an individual failure and will not affect subsequent Run/Verify blocks.

pohly replied:
Showing multiple failures at once is the better default. We can later add a `--fail-fast` boolean or `--maximum-errors` value.

pohly commented on test/checkoutput/main.go:
```diff
@@ -0,0 +1,301 @@
+	// TODO: Currently we capture the output of just the stdout. Logs extensively write to
+	// stderr, we need to come up with the right logic for validation
+	if string(output) == rv.verify {
+		fmt.Printf("Validation Success: command : %v \n Matched text: %v\n", rv.run, rv.verify)
```
All of the fmt.Printf calls should be `klog.InfoS` calls, potentially with `V()` if less important.

pchan replied:
I had pushed a commit that did not have the change. The latest code has the klogs.

pchan replied:
Fixed.

pohly commented on test/checkoutput/sample_input.md:
````diff
@@ -0,0 +1,80 @@
+
+Run:
+```console
+go run /home/prasadc/dev/k8s/repo/k8s/k1/src/k8s.io/kubernetes/staging/src/k8s.io/component-base/logs/example/cmd/logger.go --logging-format json
````
I still see the `/home/prasadc/dev/...` path here.

As this is now inside `test/checkoutput`, you probably need to use `../../staging/k8s.io/kubernetes/staging/src/k8s.io/component-base/logs/example/cmd/logger.go`.

However, when I try `go run .` inside this directory, the tool gets stuck:
```
test/checkoutput$ go run .
v log level is  1 args is  [.]
Processing root directory  .
Successful Main package Lookup: path name is  k8s.io/kubernetes/test/checkoutput
Parsing mdfile  README.md  in dir  ./.
^Csignal: interrupt
```

Running it under a debugger I see that it invokes `go run .` as the command whose output it captures, not the one specified here. Therefore it gets stuck in an endless recursion.

Logging the commands that it invokes would be very useful here.

pchan replied:
Fixed .

pohly reviewed (commented):
I had another look, but it seems that this really is still WIP and not working yet.

dashpole commented:
/assign pohly
/triage accepted

pohly commented on test/checkoutput/main.go:
```diff
@@ -95,11 +92,11 @@ func (rv *runVerify) validate() error {
 	// stderr, we need to come up with the right logic for validation
 	if string(output) == rv.verify {
-		fmt.Printf("Validation Success: command : %v \n Matched text: %v\n", rv.run, rv.verify)
+		klog.V(2).Infof("Validation Success: command : %v \n Matched text: %v\n", rv.run, rv.verify)
```
We are from the WG Structured Logging - please don't use unstructured logging calls :grin:

pchan commented:
Based on the discussion in the last structured logging meeting, the idea is to introduce a mechansim by which the user can specify how to ignore log timestamp. One option is to introduce a new optional section called metadata (similar to 'Run' and 'Expected Output') where the user can specify regex based filtering to ignore headers. I propose 2 fields:

- custom_regex_ignore
- custom_json_ignore

Optionally, to make it easy for consumers, we can add certain keywords like "ignore_klog_headers", which the tool will map to an internally maintained regex, that will be used for log comparison. 

Example

> Metadata:
> ignore_klog_headers, custom_regex:^[IWEF][0-9]{4}\\s[0-9]{2}:[0-9]{2}:[0-9]{2}\\.[0-9]{6}\\s*[0-9]*\\s.*\\.go:[0-9]{2}],custom_json_ignore:ts

If it is a json output, we need to drop the timestamp (specified above as "ts") parameter. I have currently pushed a change that compares the current klog regex, in the next iteration I will push the above feature. Please let me know your thoughts on this.

pohly commented:
> I propose 2 fields:
> 
> *  custom_regex_ignore
> *  custom_json_ignore
> 
> Optionally, to make it easy for consumers, we can add certain keywords like "ignore_klog_headers", which the tool will map to an internally maintained regex, that will be used for log comparison.

That looks reasonable to me. Just one suggestion: let's drop the `custom_` prefix. IMHO it doesn't add anything in terms of clarifying the values.

pohly commented:
One thing that needs to be clarified: is output being compare line-by-line or as a single string?

`^[IWEF]` only works when comparing line-by-line, otherwise `^` doesn't match in the middle. This can be fixed by using `(?m)^[IWEF]`.

I prefer to match against the entire output. There may be the need to ignore text only after some text on the previous line.

pchan commented:
> That looks reasonable to me. Just one suggestion: let's drop the `custom_` prefix. IMHO it doesn't add anything in terms of clarifying the values.

Yes, I agree and I will the drop the custom prefix.

pchan commented:
> One thing that needs to be clarified: is output being compare line-by-line or as a single string?
> 
> `^[IWEF]` only works when comparing line-by-line, otherwise `^` doesn't match in the middle. This can be fixed by using `(?m)^[IWEF]`.
> 
> I prefer to match against the entire output. There may be the need to ignore text only after some text on the previous line.

Yes, in the current version I have used line-by-line comparison. I found that I can use the multiline flag and named groups to extract the entire text using the regex.FindAllStringSubmatch API. 

The regex now looks like this:

`regexp.MustCompile("(?m)(?:^[IWEF][0-9]{4}\\s[0-9]{2}:[0-9]{2}:[0-9]{2}\\.[0-9]{6}\\s*[0-9]*\\s.*\\.go:[0-9]{2}])(.*)")`

Problem is when I use FindAllStringSubmatch, I still need to iterate through the matches, append them to a string and then do the compare. So practically not much different from line-by-line comparison. Please let me know if you have a way of extracting just the text and comparing.

pohly commented:
> Problem is when I use FindAllStringSubmatch, I still need to iterate through the matches, append them to a string and then do the compare.

Isn't it the other way around? Any text matching the expression(s) needs to be removed.

`regexp.Regexp.Split` could be useful here, with `n=-1`. Join the returned `[]string` slice with `strings.Join(substrings, "")` and the result can be compared.

pohly commented:
One possible enhancement would be to replace matches with a fixed string explaining why the text was removed:

```
```
I0605 22:03:07.224293 3228948 logger.go:58] Log using Infof, key: value
I0605 22:03:07.224378 3228948 logger.go:59] "Log using InfoS" key="value"
E0605 22:03:07.224393 3228948 logger.go:61] Log using Errorf, err: fail
E0605 22:03:07.224402 3228948 logger.go:62] "Log using ErrorS" err="fail"
```

=>

```
```
<<klog header>> Log using Infof, key: value
<<klog header>> "Log using InfoS" key="value"
<<klog header>> Log using Errorf, err: fail
<<klog header>> "Log using ErrorS" err="fail"
```

That may help someone not familiar with the tool understand why a test failure report looks different from the original output. It may also help debug issues with regexp matching.

pchan commented:
> > Problem is when I use FindAllStringSubmatch, I still need to iterate through the matches, append them to a string and then do the compare.
> 
> Isn't it the other way around? Any text matching the expression(s) needs to be removed.
> 
> `regexp.Regexp.Split` could be useful here, with `n=-1`. Join the returned `[]string` slice with `strings.Join(substrings, "")` and the result can be compared.

The above is possible if the match is simple like a character class but in our case we have a whole regex. The regex [syntax](https://github.com/google/re2/wiki/Syntax) does not support lookbehind (?<=). Even it supports lookbehind, I have read about issues with using regex in lookbehind. This requires us to use named groups. The result is a 2d array where each row pertains to the match of a line and column is defined by the named groups. The first result in the column is the entire match and subsequent entries correspond to named groups.

pchan commented:
> One possible enhancement would be to replace matches with a fixed string explaining why the text was removed:
> 
> That may help someone not familiar with the tool understand why a test failure report looks different from the original output. It may also help debug issues with regexp matching.

Agree, will add the fixed string in lieu of the removed text.

pohly commented:
> The above is possible if the match is simple like a character class but in our case we have a whole regex. The regex [syntax](https://github.com/google/re2/wiki/Syntax) does not support lookbehind (?<=).

Okay, so the goal is not just to replace the entire regexp, but also to keep some parts of it. Let's specify that only named groups inside the match get replaced, and that they get replace by `< name of group >`.

Then you can call https://pkg.go.dev/regexp#Regexp.FindAllSubmatchIndex. Keep all parts of the string that are:
- outside of the matches,
- inside a match but not in a group,
- matched by an unnamed group.

Then replace the text that is matched by a named group with the name of the group. https://pkg.go.dev/regexp#Regexp.SubexpNames gives you the name of the groups.

pohly reviewed (changes requested):
When I run `hack/verify-checkoutput.sh` in your branch, I get test failures:
```
/nvme/gopath/src/k8s.io/kubernetes
E0929 15:02:52.448656  189531 main.go:206] "Validation Failure" err=<
	Mismatch in Run command : go run . --logging-format json
	 processed text :
	I0929 15:02:52.443771  190489 logger.go:45] "Oops, I shouldn't be logging yet!"
	This is normal output via stdout.
	This is other output via stderr.
	{"ts":1664456572444.0156,"caller":"example/logger.go:77","msg":"Log using Infof, key: value\n","v":0}
	{"ts":1664456572444.04,"caller":"example/logger.go:78","msg":"Log using InfoS","v":0,"key":"value"}
	{"ts":1664456572444.078,"caller":"example/logger.go:80","msg":"Log using Errorf, err: fail\n"}
	{"ts":1664456572444.0928,"caller":"example/logger.go:81","msg":"Log using ErrorS","err":"fail"}
	{"ts":1664456572444.1323,"caller":"example/logger.go:83","msg":"Log with sensitive key, data: {\"secret\"}\n","v":0}
	{"ts":1664456572444.148,"caller":"example/logger.go:88","msg":"Now the default logger is set, but using the one from the context is still better.","v":0}
	{"ts":1664456572444.1648,"caller":"example/logger.go:91","msg":"Log sensitive data through context","v":0,"data":{"key":"secret"}}
	{"ts":1664456572444.2126,"caller":"example/logger.go:95","msg":"runtime","v":0,"duration":"1m0s"}
	{"ts":1664456572444.231,"caller":"example/logger.go:96","msg":"another runtime","v":0,"duration":"1h0m0s","duration":"1m0s"}
	 
	expected text :
	I0329 11:38:01.782592   99945 logger.go:44] "Oops, I shouldn't be logging yet!"
	This is normal output via stdout.
	This is other output via stderr.
	{"ts":1648546681782.9036,"caller":"cmd/logger.go:76","msg":"Log using Infof, key: value\n","v":0}
	{"ts":1648546681782.9392,"caller":"cmd/logger.go:77","msg":"Log using InfoS","v":0,"key":"value"}
	{"ts":1648546681782.9763,"caller":"cmd/logger.go:79","msg":"Log using Errorf, err: fail\n"}
	{"ts":1648546681782.9915,"caller":"cmd/logger.go:80","msg":"Log using ErrorS","err":"fail"}
	{"ts":1648546681783.0208,"caller":"cmd/logger.go:82","msg":"Log with sensitive key, data: {\"secret\"}\n","v":0}
	{"ts":1648546681783.0364,"caller":"cmd/logger.go:87","msg":"Now the default logger is set, but using the one from the context is still better.","v":0}
	{"ts":1648546681783.0552,"caller":"cmd/logger.go:90","msg":"Log sensitive data through context","v":0,"data":{"key":"secret"}}
	{"ts":1648546681783.1091,"caller":"cmd/logger.go:94","msg":"runtime","v":0,"duration":"1m0s"}
	{"ts":1648546681783.1257,"caller":"cmd/logger.go:95","msg":"another runtime","v":0,"duration":"1h0m0s","duration":"1m0s"}
 >
E0929 15:02:53.348586  189531 main.go:206] "Validation Failure" err=<
	Mismatch in Run command : go run .
	 processed text :
	E0929 15:02:53.344331  190846 main.go:321] "State transition" err=<
		Cannot call checkoutput tool in validate
	 >
	E0929 15:02:53.344533  190846 main.go:321] "State transition" err=<
		Cannot call checkoutput tool in validate
	 >
	E0929 15:02:53.344623  190846 main.go:321] "State transition" err=<
		Cannot call checkoutput tool in validate
	 >
	 
	expected text :
	klog_header Log using Infof, key: value
	klog_header "Log using InfoS" key="value"
	klog_header Log using Errorf, err: fail
	klog_header "Log using ErrorS" err="fail"
	klog_header Log message has been redacted. Log argument #0 contains: [secret-key]
 >
E0929 15:02:53.648364  189531 main.go:206] "Validation Failure" err=<
	Mismatch in Run command : go run .
	 processed text :
	E0929 15:02:53.644062  190978 main.go:321] "State transition" err=<
		Cannot call checkoutput tool in validate
	 >
	E0929 15:02:53.644276  190978 main.go:321] "State transition" err=<
		Cannot call checkoutput tool in validate
	 >
	E0929 15:02:53.644330  190978 main.go:321] "State transition" err=<
		Cannot call checkoutput tool in validate
	 >
	 
	expected text :
	klog_header Log using Infof, key: value
	klog_header "Log using InfoS" key="value"
	klog_header Log using Errorf, err: fail
	klog_header "Log using ErrorS" err="fail"
	klog_header Log message has been redacted. Log argument #0 contains: [secret-key]
 >
```

Should the script still fail? It looks like 

Some comments:
- Why does it print `/nvme/gopath/src/k8s.io/kubernetes`? It's checking `staging/src/k8s.io/component-base/logs/example/README.md`.
- Instead of the directory that is being checked, let's print the file name that is being processed. We can have more than one per directory, can't we?
- Test failures are normal output of the script and should go to stdout, without using klog. That gives us full control over the output.
- Let's print a diff between expected and actual output. In particular when just a few lines in long output are wrong that'll be more readable. You can use https://pkg.go.dev/github.com/google/go-cmp/cmp#Diff for that. It's overkill for plain strings, but that's okay.

pchan commented:
> When I run `hack/verify-checkoutput.sh` in your branch, I get test failures:
> Should the script still fail? It looks like

Yes it's supposed to fail, because the block that fails does not have the Metadata field ([link to file](https://github.com/kubernetes/kubernetes/blob/4d5d2c522f5b973ca764d220b8d1cca3dd3b65de/staging/src/k8s.io/component-base/logs/example/README.md)). I used that block as an example for Validation failure.

 
> Some comments:
> 
> * Why does it print `/nvme/gopath/src/k8s.io/kubernetes`? It's checking `staging/src/k8s.io/component-base/logs/example/README.md`.

Will fix it and print filename.

> * Instead of the directory that is being checked, let's print the file name that is being processed. We can have more than one per directory, can't we?

It's printed when verbosity is set to 2.  See example below when spawned with (v 2)
>./hack/verify-checkoutput.sh -v 2 staging/src/k8s.io/component-base/logs/
```
I0929 06:35:11.571120 3716671 main.go:184] "==================New Run====================" markdown-file="staging/src/k8s.io/component-base/logs/example/README.md"
I0929 06:35:11.571139 3716671 main.go:185] "Binary related information" binary="go" arguments=[run . --logging-format json] directory="staging/src/k8s.io/component-base/logs/example" run-verify directory="staging/src/k8s.io/component-base/logs/example" exec-name="main"
I0929 06:35:11.920634 3716671 main.go:149] "Ignore Headers" regexes=[]
E0929 06:35:11.920686 3716671 main.go:206] "Validation Failure" err=<
```

> * Test failures are normal output of the script and should go to stdout, without using klog. That gives us full control over the output.

I will update it.

> * Let's print a diff between expected and actual output. In particular when just a few lines in long output are wrong that'll be more readable. You can use https://pkg.go.dev/github.com/google/go-cmp/cmp#Diff for that. It's overkill for plain strings, but that's okay.

I will update to do a diff and print.

pohly commented:
> Yes it's supposed to fail, because the block that fails does not have the Metadata field. I used that block as an example for Validation failure.

Let's make it a goal that the verify script doesn't fail.

We then need a different way of testing that the tool works. Can that be done in a _test.go file which contains different test scenarios? It could set up a tmp directory with a certain set of files as specified for the test scenario, then run the tool and check the output of the tool itself. It becomes a bit recursive, though - we need the tool to check itself :grin:

dims commented:
If you still need this PR then please rebase, if not, please close the PR

pchan committed:
Add unit-testing. Fix inconsistencies.

pchan commented on test/checkoutput/README.md:
```diff
@@ -0,0 +1,70 @@
+
+### Metadata
+
+The MetaData field is nominally optional, but will need to be specified to process out the timestamps and location based information. Each line in the block is either a regex or a keyword that will be applied to remove headers in the log. It needs to be specified in the following format.
```
Only the current test-case.

pchan committed:
Fix comment

pchan commented:
I have added unit-testing for the tool. Please review.

mengjiao-liu commented:
Do you still work for it? If so, please rebase this PR.

dims commented:
closing, please rebase and reopen when ready.
